File: scripts/scrape-whisky.js
import { createClient } from '@supabase/supabase-js'
import https from 'https'
import http from 'http'
import { URL } from 'url'

const SUPABASE_URL         = process.env.SUPABASE_URL
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY
const GEMINI_KEY           = process.env.GEMINI_KEY

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  console.error('[error] Missing SUPABASE_URL or SUPABASE_SERVICE_KEY')
  process.exit(1)
}

const sb = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY)
const GEMMA_MODEL = 'gemma-3-27b-it'
const GEMMA_URL   = `https://generativelanguage.googleapis.com/v1beta/models/${GEMMA_MODEL}:generateContent?key=${GEMINI_KEY}`

function httpGet(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url)
    const protocol  = parsedUrl.protocol === 'https:' ? https : http
    const opts      = {
      hostname: parsedUrl.hostname,
      path:     parsedUrl.pathname + parsedUrl.search,
      method:   'GET',
      headers:  { 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36', ...headers },
    }
    const req = protocol.request(opts, res => {
      const chunks = []
      res.on('data', c => chunks.push(c))
      res.on('end',  () => resolve(Buffer.concat(chunks).toString('utf8')))
    })
    req.on('error', reject)
    req.setTimeout(15000, () => { req.destroy(); reject(new Error('Request timeout')) })
    req.end()
  })
}

function textBetween(html, start, end) {
  const idx = html.indexOf(start)
  if (idx === -1) return ''
  const from = idx + start.length
  const stop = html.indexOf(end, from)
  return stop === -1 ? '' : html.slice(from, stop).trim()
}

function getMetaContent(html, property) {
  const re = new RegExp(`<meta[^>]+(?:property|itemprop|name)=["']${property}["'][^>]+content=["']([^"']+)["']`, 'i')
  const m  = html.match(re)
  if (m) return m[1]
  const re2 = new RegExp(`<meta[^>]+content=["']([^"']+)["'][^>]+(?:property|itemprop|name)=["']${property}["']`, 'i')
  const m2 = html.match(re2)
  return m2 ? m2[1] : ''
}

async function scrape(url) {
  console.log(`[scraper] Fetching: ${url}`)
  const html = await httpGet(url)

  const name      = textBetween(html, '"og:title" content="', '"') ||
                    textBetween(html, '<title>', '</title>').split('|')[0].trim()
  const distillery = textBetween(html, '"og:title" content="', '"')
    ?.replace('Whisky Review:', '').replace('Whisky:', '').split('-')[0].trim() || ''
  const description = textBetween(html, 'itemprop="description" content="', '"') ||
                      textBetween(html, '"description" content="', '"')
  const priceStr = textBetween(html, 'itemprop="price" content="', '"') ||
                   textBetween(html, 'class="price"', '<').replace(/[^0-9.,]/g, '').trim()

  const abv       = textBetween(html, 'ABV:', '<').replace(/[^0-9.]/g, '') ||
                    textBetween(html, '"abv"', '"').replace(/[^0-9.]/g, '')
  const age       = textBetween(html, 'Age:', '<').replace(/[^0-9]/g, '') ||
                    textBetween(html, '"age"', '"').replace(/[^0-9]/g, '')
  const region    = textBetween(html, 'Region:', '<').trim() ||
                    textBetween(html, '"region"', '"').trim()
  const country   = textBetween(html, 'Country:', '<').trim() ||
                    getMetaContent(html, 'og:locale').split('_')[1] ||
                    textBetween(html, '"country"', '"').trim()

  let type = 'other'
  const lower = (name + ' ' + region + ' ' + description).toLowerCase()
  if (lower.includes('japanese') || lower.includes('nippon') || country === 'Japan') type = 'japanese'
  else if (lower.includes('scotch') || lower.includes('speyside') || lower.includes('islay') ||
           lower.includes('highland') || lower.includes('lowland') || region.toLowerCase().includes('scotland')) type = 'scotch'
  else if (lower.includes('bourbon') || lower.includes('tennessee') || country === 'United States') type = 'bourbon'
  else if (lower.includes('irish') || country === 'Ireland') type = 'irish'

  let price_band = null
  if (priceStr) {
    const num = parseFloat(priceStr.replace(',', '.'))
    if (!isNaN(num)) {
      if (num < 40)       price_band = 'budget'
      else if (num < 80)  price_band = 'mid-range'
      else if (num < 150) price_band = 'premium'
      else if (num < 300) price_band = 'luxury'
      else                 price_band = 'super-premium'
    }
  }

  return {
    name:        name.replace(/(Whisky Review|Whisky:|—)/g, '').trim(),
    distillery,
    country:     country || null,
    region:      region  || null,
    type,
    age:         age     ? parseInt(age) : null,
    abv:         abv     ? parseFloat(abv) : null,
    price_band,
    status:      null,
  }
}

async function callGemma(prompt) {
  if (!GEMINI_KEY) return null
  try {
    const res = await fetch(GEMMA_URL, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        generationConfig: { temperature: 0.3, maxOutputTokens: 300, responseMimeType: 'application/json' },
      }),
    })
    if (!res.ok) return null
    const data  = await res.json()
    const text  = data.candidates?.[0]?.content?.parts?.[0]?.text || ''
    if (!text)  return null
    const match = text.match(/\{[\s\S]*\}/)
    return match ? JSON.parse(match[0]) : null
  } catch {
    return null
  }
}

function parseFlavorResponse(parsed) {
  if (!parsed) return {}
  const clamp = v => Math.min(5, Math.max(0, Math.round(Number(v) || 0)))
  return {
    nose:      (parsed.nose      || '').trim(),
    palate:    (parsed.palate    || '').trim(),
    dulzor:    clamp(parsed.dulzor),
    ahumado:   clamp(parsed.ahumado),
    cuerpo:    clamp(parsed.cuerpo),
    frutado:   clamp(parsed.frutado),
    especiado: clamp(parsed.especiado),
  }
}

async function main() {
  const url = process.argv[2]
  if (!url) {
    console.error('Usage: node scripts/scrape-whisky.js <distilld-url>')
    process.exit(1)
  }

  const data = await scrape(url)
  console.log('[scraper] Parsed whisky:', data.name)
  console.log('[scraper] Fields:', JSON.stringify(data, null, 2))

  const { data: existing } = await sb
    .from('catalogue')
    .select('id, name')
    .ilike('name', data.name)
    .limit(1)

  if (existing?.length) {
    console.log(`[scraper] Already exists in catalogue: "${existing[0].name}" (id=${existing[0].id})`)
    process.exit(0)
  }

  if (GEMINI_KEY) {
    const parts = [data.name]
    if (data.distillery) parts.push(`by ${data.distillery}`)
    if (data.country)    parts.push(`from ${data.country}`)
    if (data.region)     parts.push(`(${data.region})`)
    if (data.age)        parts.push(`aged ${data.age}`)
    if (data.abv)        parts.push(`at ${data.abv}%`)
    if (data.type)       parts.push(`style: ${data.type}`)

    const prompt = `You are an expert whisky taster. Provide tasting notes and flavor scores for:
${parts.join(', ')}.

Respond ONLY with a JSON object, no markdown, no explanation:
{
  "nose": "2-4 aroma descriptors, comma separated",
  "palate": "2-4 taste descriptors, comma separated",
  "dulzor": <integer 0-5, sweetness>,
  "ahumado": <integer 0-5, smokiness>,
  "cuerpo": <integer 0-5, body/weight>,
  "frutado": <integer 0-5, fruitiness>,
  "especiado": <integer 0-5, spiciness>
}`

    console.log('[scraper] Calling Gemma for flavor profile…')
    const flavor = await callGemma(prompt)
    if (flavor) {
      Object.assign(data, parseFlavorResponse(flavor))
      data.status = true
      console.log('[scraper] Flavor profile generated:', JSON.stringify(data, null, 2))
    }
  }

  const { data: inserted, error } = await sb
    .from('catalogue')
    .insert([data])
    .select()
    .single()

  if (error) {
    console.error('[scraper] Insert failed:', error.message)
    process.exit(1)
  }

  console.log(`[scraper] ✓ Added to catalogue: "${inserted.name}" (id=${inserted.id})`)
}

main().catch(err => { console.error('[scraper] Error:', err.message); process.exit(1) })

File: scripts/scrape_whisky.py
#!/usr/bin/env python3
import os
import re
import sys
import json
import urllib.request
import urllib.error
from html import unescape
from urllib.parse import urlparse
from pathlib import Path

def load_env():
    env_path = Path(__file__).parent.parent / '.env'
    if env_path.exists():
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    os.environ.setdefault(key, value)

load_env()

SUPABASE_URL = os.environ.get('SUPABASE_URL') or os.environ.get('VITE_SUPABASE_URL')
SUPABASE_SERVICE_KEY = os.environ.get('SUPABASE_SERVICE_KEY')
GEMINI_KEY = os.environ.get('GEMINI_KEY')

if not SUPABASE_URL or not SUPABASE_SERVICE_KEY:
    print('[error] Missing SUPABASE_URL or SUPABASE_SERVICE_KEY')
    print('Make sure .env file exists with SUPABASE_SERVICE_KEY and VITE_SUPABASE_URL')
    sys.exit(1)


def http_get(url):
    req = urllib.request.Request(
        url,
        headers={'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'}
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            return response.read().decode('utf-8')
    except urllib.error.URLError as e:
        raise Exception(f'Request failed: {e}')


def attr(html, attr_name):
    pattern = rf'<meta[^>]+\s{re.escape(attr_name)}=["\']([^"\']+)["\']'
    match = re.search(pattern, html, re.IGNORECASE)
    return match.group(1) if match else ''


def decode_entities(text):
    return unescape(text)


def extract_fact(html, label):
    pattern = rf'{label}[\s\S]{{0,120}}?<p[^>]*class="h3"[^>]*>([^<]+)</p>'
    match = re.search(pattern, html, re.IGNORECASE)
    return decode_entities(match.group(1).strip()) if match else ''


def extract_distillery(html):
    pattern = r'Distillery[\s\S]{0,120}?<p[^>]*class="h3"[^>]*>([^<]+)</p>'
    match = re.search(pattern, html, re.IGNORECASE)
    return decode_entities(match.group(1).strip()) if match else ''


def extract_all_facts(html):
    pattern = r'<p class="h3"[^>]*>([^<]+)</p>'
    matches = re.findall(pattern, html)
    return [decode_entities(m.strip()) for m in matches]


def scrape(url):
    print(f'[scraper] Fetching: {url}')
    html = http_get(url)

    raw_name = attr(html, 'property="og:title"') or ''
    if not raw_name:
        title_match = re.search(r'<title>([^<]+)', html)
        if title_match:
            raw_name = title_match.group(1).split('|')[0].strip()
    name = decode_entities(raw_name)

    facts = extract_all_facts(html)
    distillery_from_facts = None
    for i, v in enumerate(facts):
        if i > 0 and v not in ('Yes', 'No', 'Single Malt', 'Blended', 'Grain'):
            distillery_from_facts = v
            break

    abv = extract_fact(html, 'ABV').replace('%', '')
    if not abv:
        abv_match = re.search(r'ABV[\s\S]{0,120}?<p[^>]*class="h3"[^>]*>(\d+)', html)
        if abv_match:
            abv = abv_match.group(1)

    country = extract_fact(html, 'Country') or ''
    type_str = extract_fact(html, 'Type') or ''
    distillery = extract_distillery(html) or distillery_from_facts or ''
    region = extract_fact(html, 'Region') or ''
    age_str = extract_fact(html, 'Age') or ''
    price_match = re.search(r'€(\d+)', html)
    price_str = price_match.group(1) if price_match else ''

    raw_locale = attr(html, 'property="og:locale"')
    locale_country = raw_locale.split('_').pop() if raw_locale else ''
    if not country and locale_country:
        locale_map = {'gb': 'United Kingdom', 'us': 'United States', 'jp': 'Japan', 'ie': 'Ireland'}
        country = locale_map.get(locale_country.lower(), locale_country)

    whisky_type = 'other'
    lower = (name + ' ' + country + ' ' + type_str).lower()
    if 'japanese' in lower or 'nippon' in lower or country == 'Japan':
        whisky_type = 'japanese'
    elif any(x in lower for x in ['scotch', 'speyside', 'islay', 'highland', 'lowland']):
        whisky_type = 'scotch'
    elif 'bourbon' in lower or 'tennessee' in lower or country == 'United States':
        whisky_type = 'bourbon'
    elif 'irish' in lower or country == 'Ireland':
        whisky_type = 'irish'
    elif type_str and 'blended' in type_str.lower():
        whisky_type = 'japanese'

    distillery_raw = distillery or None
    if not distillery_raw:
        known_distilleries = ['Hibiki', 'Yamazaki', 'Hakushu', 'Nikka', 'Macallan', 'Glenfiddich', 'Talisker', 'Lagavulin']
        for d in known_distilleries:
            if d in name:
                distillery_raw = d
                break

    price_band = None
    if price_str:
        try:
            num = float(price_str.replace(',', '.'))
            if num < 40:
                price_band = 'budget'
            elif num < 80:
                price_band = 'mid-range'
            elif num < 150:
                price_band = 'premium'
            elif num < 300:
                price_band = 'luxury'
            else:
                price_band = 'super-premium'
        except ValueError:
            pass

    print(f'[scraper] Parsed: name="{name}" distillery="{distillery}" country="{country}" type="{whisky_type}" abv="{abv}" price="{price_str}" price_band="{price_band}"')
    print(f'[scraper] Facts: [{", ".join(facts)}]')

    name_clean = re.sub(r'(Whisky Review|Whisky:|—)', '', name).strip()

    return {
        'name': name_clean,
        'distillery': distillery_raw,
        'country': country or None,
        'region': region or None,
        'type': whisky_type,
        'age': int(age_str) if age_str.isdigit() else None,
        'abv': float(abv) if abv else None,
        'price_band': price_band,
    }


def call_gemma(prompt):
    if not GEMINI_KEY:
        return None

    model = 'gemma-3-27b-it'
    url = f'https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={GEMINI_KEY}'

    payload = {
        'contents': [{'role': 'user', 'parts': [{'text': prompt}]}],
        'generationConfig': {'temperature': 0.3, 'maxOutputTokens': 300, 'responseMimeType': 'application/json'}
    }

    try:
        req = urllib.request.Request(
            url,
            data=json.dumps(payload).encode('utf-8'),
            headers={'Content-Type': 'application/json'}
        )
        with urllib.request.urlopen(req, timeout=30) as response:
            data = json.loads(response.read().decode('utf-8'))
            text = data.get('candidates', [{}])[0].get('content', {}).get('parts', [{}])[0].get('text', '')
            if text:
                match = re.search(r'\{[\s\S]*\}', text)
                return json.loads(match.group(0)) if match else None
    except Exception as e:
        print(f'[gemma] Error: {e}')
    return None


def parse_flavor_response(parsed):
    if not parsed:
        return {}

    def clamp(v):
        return min(5, max(0, round(float(v or 0))))

    return {
        'nose': (parsed.get('nose') or '').strip(),
        'palate': (parsed.get('palate') or '').strip(),
        'dulzor': clamp(parsed.get('dulzor')),
        'ahumado': clamp(parsed.get('ahumado')),
        'cuerpo': clamp(parsed.get('cuerpo')),
        'frutado': clamp(parsed.get('frutado')),
        'especiado': clamp(parsed.get('especiado')),
    }


def supabase_insert(data):
    url = f'{SUPABASE_URL}/rest/v1/catalogue'
    headers = {
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': f'Bearer {SUPABASE_SERVICE_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation'
    }

    payload = json.dumps([data])

    req = urllib.request.Request(url, data=payload.encode('utf-8'), headers=headers, method='POST')
    try:
        with urllib.request.urlopen(req) as response:
            result = json.loads(response.read().decode('utf-8'))
            return result
    except urllib.error.HTTPError as e:
        error_body = e.read().decode('utf-8')
        raise Exception(f'Insert failed: {e.code} - {error_body}')


def check_existing(name):
    url = f'{SUPABASE_URL}/rest/v1/catalogue?name=ilike.*%25{urllib.parse.quote(name)}%25&limit=1'
    headers = {
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': f'Bearer {SUPABASE_SERVICE_KEY}'
    }

    req = urllib.request.Request(url, headers=headers)
    with urllib.request.urlopen(req) as response:
        return json.loads(response.read().decode('utf-8'))


def main():
    if len(sys.argv) < 2:
        print('Usage: python scripts/scrape_whisky.py <distilld-url>')
        sys.exit(1)

    url = sys.argv[1]
    data = scrape(url)

    existing = check_existing(data['name'])
    if existing:
        print(f'[scraper] Already exists in catalogue: "{existing[0]["name"]}" (id={existing[0]["id"]})')
        sys.exit(0)

    if GEMINI_KEY:
        parts = [data['name']]
        if data.get('distillery'):
            parts.append(f"by {data['distillery']}")
        if data.get('country'):
            parts.append(f"from {data['country']}")
        if data.get('age'):
            parts.append(f"aged {data['age']}")
        if data.get('abv'):
            parts.append(f"at {data['abv']}%")
        if data.get('type'):
            parts.append(f"style: {data['type']}")

        prompt = f'''You are an expert whisky taster. Provide tasting notes and flavor scores for:
{", ".join(parts)}.

Respond ONLY with a JSON object, no markdown, no explanation:
{{
  "nose": "2-4 aroma descriptors, comma separated",
  "palate": "2-4 taste descriptors, comma separated",
  "dulzor": <integer 0-5, sweetness>,
  "ahumado": <integer 0-5, smokiness>,
  "cuerpo": <integer 0-5, body/weight>,
  "frutado": <integer 0-5, fruitiness>,
  "especiado": <integer 0-5, spiciness>
}}'''

        print('[scraper] Calling Gemma for flavor profile…')
        flavor = call_gemma(prompt)
        if flavor:
            data.update(parse_flavor_response(flavor))
            data['status'] = True

    result = supabase_insert(data)
    print(f'[scraper] ✓ Added to catalogue: "{data["name"]}"')


if __name__ == '__main__':
    main()